Add integration test for IQv2 in KIP-1071

bbejeck opened the pull request:
Adds a parameterized integration test ensuring that the same host->partitions information is sent in both `classic` and `streams` group mode.

This is a stacked PR only review the last commit.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
Merging the Refactory to add standby tasks heartbeat response
added test

Add stability check and refactor partition assignments

Introduces a `membersStable` method in `StreamsGroup` to ensure group members are in a stable state before partition assignments. Refactors `EndpointToPartitionsManager` to streamline input topic handling and updates related tests for clarity and correctness. Adds integration test for validating partition assignment logic with multiple configurations.

bbejeck commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:
```diff
@@ -2382,17 +2382,19 @@ private CoordinatorResult<StreamsGroupHeartbeatResult, CoordinatorRecord> stream
-                responseEndpoint.setPort(endpoint.port());
-                StreamsGroupHeartbeatResponseData.EndpointToPartitions endpointToPartitions = EndpointToPartitionsManager.endpointToPartitions(groupMember, responseEndpoint, group);
-                endpointToPartitionsList.add(endpointToPartitions);
+        if (group.membersStable()) {
```
Adding this was required - otherwise we could end up trying to set up the `EndpointToPartitions` when a new member assignment is not finalized.  I discovered this developing the integration test - the newly added member was in a `UNRELEASED_TASKS`.  I'm open to suggestions if anyone doesn't agree with my approach here.

bbejeck commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -454,6 +454,15 @@ public Map<String, StreamsGroupMember> members() {
         return Collections.unmodifiableMap(members);
     }
 
+    public boolean membersStable() {
```
Added this method - explained why above

bbejeck commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/topics/EndpointToPartitionsManager.java:
```diff
@@ -47,11 +46,11 @@ public static StreamsGroupHeartbeatResponseData.EndpointToPartitions endpointToP
             final Map<String, TopicMetadata> groupTopicMetadata = streamsGroup.partitionMetadata();
             Set<Map.Entry<String, Set<Integer>>> taskEntrySet = streamsGroupMember.assignedActiveTasks().entrySet();
-            List<StreamsGroupHeartbeatResponseData.TopicPartition> topicPartitionList = getTopicPartitions(taskEntrySet, sourceTopics, repartitionSourceTopics, groupTopicMetadata);
+            List<StreamsGroupHeartbeatResponseData.TopicPartition> topicPartitionList = getTopicPartitions(taskEntrySet, sourceTopics, groupTopicMetadata);
```
I was incorrectly combining source and repartition topics - they should be added separately since we are sending back standby tasks separately

lucasbru replied:
It's good that you wrote an integration test, so I can trust that you are doing the right thing here. I just don't really understand why IQv2 works this way. Why do we only send input topics for active taskss and repartition topics for standby tasks?

bbejeck replied:
@lucasbru - good question.  I'll update the test to be more clear, but it should be the case that active tasks for both instances get source partitions.  Since there's two sub-topologies with two partitions, each instance gets one of each.

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2EndpointToPartitionsIntegrationTest.java:
```diff
@@ -0,0 +1,226 @@
+        cluster.start();
+    }
+
+    public void setUp(final TestInfo testInfo) throws InterruptedException {
```
This is required so we can parameterize the `EmbeddedKafkaCluster` to start with a configuration of streams standby tasks.  Once we have implemented the broker side acceptance of config requests, we can go back to a more standard test setup.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/utils/TestUtils.java:
```diff
@@ -90,7 +90,7 @@ public static String safeUniqueTestName(final TestInfo testInfo) {
     }
 
-    private static String safeUniqueTestName(final String testName) {
+    public static String safeUniqueTestName(final String testName) {
```
Needed this since I need to generate unique test names without the `@BeforEach` annotation which passes in the `TestInfo` object.  Is there any reason this couldn't be public anyway?

bbejeck committed:
Refactor IQv2 metadata handling and endpoint partition mapping.

Streamlined metadata validation logic in IQv2 integration tests and removed redundant code. Simplified `maybeBuildEndpointToPartitions` by eliminating unnecessary stability checks, improving clarity and maintainability of endpoint-to-partition mapping.

bbejeck committed:
Remove unused `membersStable` method from StreamsGroup

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2EndpointToPartitionsIntegrationTest.java:
```diff
@@ -0,0 +1,229 @@
+                            IntegrationTestUtils.DEFAULT_TIMEOUT,
+                            () -> "Kafka Streams one or two never transitioned to a RUNNING state.");
+
+                    waitForCondition(() ->  {
```
Although we can infer the task assignments from the metadata, I wanted to have an explicit check here for the expected task movement.

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2EndpointToPartitionsIntegrationTest.java:
```diff
@@ -0,0 +1,229 @@
+                            "KafkaStreams one never released active tasks and received standby task");
+
+                    waitForCondition(() -> {
+                        final ThreadMetadata threadMetadata = streamsTwo.metadataForLocalThreads().iterator().next();
```
same here

lucasbru commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/IQv2EndpointToPartitionsIntegrationTest.java:
```diff
@@ -0,0 +1,229 @@
+                    final Set<TopicPartition> streamsTwoTopicPartitions = streamsOneMetadataTwo.topicPartitions();
+                    assertEquals(3030, streamsOneMetadataTwo.hostInfo().port());
+                    assertEquals(2, streamsTwoTopicPartitions.size());
+                    assertEquals(expectedStandbyTopicPartitionCount, streamsOneMetadataTwo.standbyTopicPartitions().size());
```
Should we check the contents of the standbyTopicPartitions? The javadoc states that they are changelog partitions, but it looks like we are not using changelog topics (up in the implementation). I am utterly confused now which topic partitions are included in which set now.

Can we also check the result of `standbyStateStoreNames`? It seems it is derived from `standbyTopicPartitions`, it should run the same logic in both protocols.

bbejeck replied:
Yes, I can update the test to verify `standbyStateStoreNames`.  

> The javadoc states that they are changelog partitions

I believe the javadoc is incorrect. I think confusion lies in the fact that standby tasks will **_consume_** from changelog topic partitions (via the restore consumer), but for IQ from standbys you'll need to query the input topic partitions (repartition ones if there's been a key changing operation for the stateful operation).  What's critical is that subtopology partitions and partitions of the changelog topics match, although I'm not sure if there's a practical way to compare those in a test - but I'll look into.

I'll do a PR to update the javadoc.

bbejeck replied:
> It seems it is derived from standbyTopicPartitions

I glossed over this before, but I don't believe that's the case - where do you see that?

lucasbru replied:
> > It seems it is derived from standbyTopicPartitions
> 
> I glossed over this before, but I don't believe that's the case - where do you see that?

https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsMetadataState.java#L392 I was looking at this - it seems like it's derived from `standbyPartitionHostMap`

lucasbru reviewed (commented):
I have some conceptual qeustions, but on the whole it's great that we are adding an integration test for this

bbejeck committed:
Refactor and enhance IQv2 test metadata validation.

Replaced redundant variable names and improved metadata handling for active and standby tasks in the IQv2 test. Added validations for state store names using `EXPECTED_STORE_NAME` and organized metadata assertions to enhance clarity and testing coverage. These changes improve maintainability and ensure better validation of stream metadata.

bbejeck committed:
Add inspection of standby topics

bbejeck commented:
@lucasbru I've updated for you comments and it's ready for another review.  I'll need to dig in some to see why we only get repartition topics for the standby tasks.  My guess that while a standby is in standby mode, it only needs the repartition topic so it can determine the downstream store and changelog topics etc.

lucasbru reviewed (approved):
LGTM, thanks!
